File: LeetCode/cpp/104 - Maximum Depth of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int rst = 0;
    int maxDepth(TreeNode *root)
    {
        if (!root)
            return 0;
        if (!root->left && !root->right)
            return 1;
        int depth = 1;
        dfs(root, depth);
        return rst;
    }

    void dfs(TreeNode *root, int depth)
    {
        if (root->left || root->right) {
            depth++;
            rst = max(rst, depth);
        }
        if (root->left)
            dfs(root->left, depth);
        if (root->right)
            dfs(root->right, depth);
    }
};

File: LeetCode/grind75/104 - Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution1:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        result = 0

        def dfs(root, depth):
            if not root:
                return

            depth += 1
            nonlocal result
            if depth > result:
                result = depth

            dfs(root.left, depth)
            dfs(root.right, depth)

        dfs(root, 0)
        return result


class Solution2:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))


class Solution3:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        q = deque()
        q.append(root)
        depth = 0

        while q:
            depth += 1

            for _ in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

        return depth
